src/System.IO.Pipelines/Pipe.cs

```diff
@@ -11,6 +11,10 @@ namespace System.IO.Pipelines
     /// </summary>
     internal class Pipe : IPipe, IPipeReader, IPipeWriter, IReadableBufferAwaiter, IWritableBufferAwaiter
     {
+        // This sync objects protects the following state:
+        // 1. _commitHead & _commitHeadIndex
```

Comment:
You need to expand the lock in AllocateWriteHead too, to cover all access to _commitHead. While it might not be technically necessary yet, it will be if/when we allow the consuming thread to return the commit head once it's caught up in order to reduce the memory utilization if idle connections. This is what socket input does. See https://github.com/aspnet/KestrelHttpServer/pull/1120